### ISSUE TITLE
[Bug] Tabs became flatten (tree structure lost)

### ISSUE THREAD

**Issue by j1elo** [piroor/treestyletab#3276]
## Abstract

Yesterday when powering off the computer, I had several TST subtrees. Today, upon opening Firefox, suddenly **one** of the subtrees had been flattened, and all its tabs are now leafs of the main tree (i.e. not nested).

I noticed https://github.com/piroor/treestyletab/issues/260 and https://github.com/piroor/treestyletab/issues/1464 but those are closed issues.

There is also "*Tabs become flatten suddenly / tree structure is lost accidentally*" in the [FAQ](https://github.com/piroor/treestyletab/blob/trunk/README.md#troubles-unexpected-behaviors), so I see this is a recurring problem and thus I'm assuming it hasn't been solved in latest development builds.

## Steps to reproduce

Given the *seemingly random* occurrence of this issue, it's not easy nor possible to reproduce in a reliable way (at least for now)

Yesterday, my tree structure looked like this:

```
A
\- B (collapsed)
   \- B1
   \- B2
   \- B3
\- C (collapsed)
   \- C1
   \- C2
   \- C3
D (selected)
```

Today, upon power-on the computer and opening Firefox, my tree structure looks like this:

```
A
B
B1
B2
B3
\- C (collapsed)
   \- C1
   \- C2
   \- C3
D (selected)
```

i.e. **one** subtree has been flattened out. Other subtrees are kept intact.

More details:

* No system or Firefox updates have occurred during this time.

    - ~I don't know if an automatic add-on update might have occurred upon opening Firefox today~ -- I just saw that current add-on version is 3.9.11, last updated December 6, 2022. So an add-on update hasn't happened between yesterday and today.

* Firefox account sync is enabled, and set to synchronize tabs.

    - The only other device in sync is an Android smartphone with Firefox for Android.
    - None of the affected subtree tabs have been opened / accessed / touched at all in the last days by either of the computer or the synchronized phone.

## Expected result

Affected subtree should still exist.

## Actual result

Affected subtree has been flattened.

## Environment

 * Platform (OS): Ubuntu 20.04
 * Version of Firefox: 108.0.2
 * Version (or revision) of Tree Style Tab: 3.9.11
 * No other tab-management add-ons are installed. Only basic Tree Style Tab, without any other extensions to it.

## Other notes

Thanks A LOT for this add-on. It's one of those things that make Firefox immensely better than alternative browsers! (among other things)

Can I do anything to help improve this issue? I see that TST has A LOT of logging options. Are there any that I should enable to aid in troubleshooting in case this happens again? (in that case, please name exactly which options I should enable)

**Comment by bhurwitz** [comment]
I have also seen this issue in the past few weeks (I'm not certain when I started noticing it), but my entire tree was flattened - somewhere in the 100-200 tabs, mostly in folder groups through the TST More Tree Commands extension, with multiple nested levels of both tabs and folders. Based on the FAQ section linked by @j1elo, I installed the Tab Session Manager (v. 6.12.1) extension and enabled TST support. Restarting the browser (and the computer) with the extension enabled did not reproduce the original flattening issue, BUT I also couldn't reproduce the issue after disabling the extension and restarting the browser (and the computer), so not clear what to think.

I'm running Firefox 108.0.2 as well, but on Win10 pro, version 22H2. Additional potentially relevant extensions include [TST] Fade old tabs (v. 1.0.1), Auto Tab Discard (v. 0.6.7), Bookmark Tree for TST (v. 4.2), Move unloaded tabs for TST (v. 2.4), Tab Unloader for TST (v. 6.15), TST Indent Line (v. 1.2.2), and TST More Tree Commands (v. 1.5).

**Comment by artonaut** [comment]
As I mentioned in the (closed) bug #1464, I have been bitten by this issue multiple times over the past year at least. I rarely restart firefox, but when I do, there is a good chance part of my tab structure is lost. In such cases, it seems that TST started restoring the tab structure, but then, at some point it stopped, and all remaining tabs in that window, and all remaining windows have their structure flattened.

I'd gladly help debugging this issue, for instance by enabling specific log options (please advise), or in another way.

Since a few months, I have started to suspect that the problem may be related to TST updates ? There seems to be a correlation between new TST versions and this problem. I have no proof yet (but see below), but could it be, that after a firefox restart and a session restore, firefox gets around to upgrading TST to its newest version while TST is busy restoring the tab hierarchy, and the upgrade causes TST to stop in its tracks, with a (partly) flattened hierarchy as a result ? Alternatively, it might also be another extension (maybe TST add-on ?) that is being upgraded, which interferes with TST's process of restoring the tab hierarchy ?

My observation from today:

Today, after an inadvertent power loss, I had the foresight to save the session backup file before starting firefox. I then started firefox as usual, and most of my tab hierarchy was gone (1st window was OK, 2nd window first few tabs OK, rest flat, all other windows flat as well). I then stopped firefox, replaced the freshly written session backup file with the one I had saved, and started firefox again. This time, my old tab hierarchy was fully restored. I am 99.99% sure that, before the restart, the active version of TST was 3.9.10, while after the restart, the active version was certainly 3.9.11.

**Comment by piroor** [comment]
I've tried to reproduce with updating of TST while the startup restoration is in-progress, but sadly it doesn't reproduce the flatten tree problem for me... What I did is:

1. Start Firefox.
2. Install TST 3.9.10.
3. Prepare tree constructed from about 1000 tabs.
4. Restart Firefox.
5. While the tab restoration is in-progress, update TST to 3.9.12.

But TST has restored the tree successfully after it was updated.

**Comment by bhurwitz** [comment]
I will say that since my last comment (3 weeks ago or so) and deliberate
restarts daily with Tabs Session Manager installed, I have not seen this
issue reappear. So that's a potential workaround for anyone else running
into this issue.
---
Ben

**Comment by artonaut** [comment]
> I've tried to reproduce with updating of TST while the startup restoration is in-progress, but sadly it doesn't reproduce the flatten tree problem for me...

OK. That's a pity :-)

Thanks for testing.

I'll gladly help in any way I can to solve this issue. I already did a number of start/stop tests with a copy of the profile that failed last time, but it did not happen again - which might be caused by the fact the profile was a copy. I assume the best option at this moment would be to enable some logging, to see what TST is doing, and hopefully catch it in the act ?

Could you recommend some logging options for me to enable in TST, that could help diagnose this issue ? Preferably, they should not generate too much data, and not have too much of a performance cost, as I will have to use them for longer periods of time (weeks, months...).

**Comment by piroor** [comment]
Note: this post just describes technical information of TST and does not provide any solution of the problem.

TST saves tree information via [the `sessions.setTabValue()` API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/sessions/setTabValue) and each tab will store its "permanent id", "ancestors" and "children". For example, if you one parent tab and two children tabs, they may have:

* **parent tab (id=1)**
  * permanent id: tab-aaa-aaa-000-000
  * ancestors: (none)
  * children:  tab-bbb-bbb-111-111, tab-ccc-ccc-222-222
* **child tab 1 (id=2)**
  * permanent id: tab-bbb-bbb-111-111
  * ancestors: tab-aaa-aaa-000-000
  * children: (none)
* **child tab 2 (id=3)**
  * permanent id: tab-ccc-ccc-222-222
  * ancestors: tab-aaa-aaa-000-000
  * children: (none)

When Firefox restores tabs and TST calls [the `sessions.getTabValue()` API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/sessions/getTabValue), Firefox returns those information based on restored session data. TST restructure tree of tabs on the next startup based on it.

So, when you get flat tabs on the next startup, it means something wrong. For example:

* Firefox didn't restore whole session data for each tab. Restored three tabs have no "permanent id", "ancestors" and "children" information so TST can't restore tree.
* Firefox didn't restore some parts of session data for each tab. Restored tree tabs have "ancestors" and "children" information but does not have restored "permanent id" - TST still cannot restore tree.
* Firefox restored all of session data but TST failed to detect "it is the time all tabs are restored and ready to restore tree". TST can restore tree but the operation is untriggered accidentally.

To know which is your case, we need to run a code to inspect session data of flattened tree. Could you try to run an inspection code with the debugger console for TST's background page, when you see a flattened tabs for a restored browser window?

1. Go to `about:debugging`.
2. Click "This Firefox" in the left pane.
3. Click "Inspect" of the "Tree Style Tab" item. Then you'll see a debugger window.
4. Click "Console" tab in the debugger.
5. Click an icon next to "..." toolbar button and choose "/background/background.html".
6. Paste this script to the prompt and run:
   ```
   browser.tabs.query({}).then(async tabs => {
     return await Promise.all(tabs.map(async tab => { return `---${tab.windowId}-${tab.id}---
   id: ${(await browser.sessions.getTabValue(tab.id, 'data-persistent-id')).id}
     ancestors: ${await browser.sessions.getTabValue(tab.id, 'ancestors') || ''}
     children:  ${await browser.sessions.getTabValue(tab.id, 'children') || ''}` }
     ))}).then(data => console.log(data.join('\n')))
   ```
7. Copy the output and paste it here.

**Comment by artonaut** [comment]
Thanks very much for this information !

I'll definitely do this next time I get a (partially) flattened tree. I am a bit hesitant about repeatedly restarting my main Firefox session without necessity though (and depending on the ultimate cause, that may even be of litte use anyway...), so it may take some time before I get back with the results.

In the mean time, I'll also try and reproduce the problem using the test profile I created. Maybe that will yield results sooner.

**Comment by artonaut** [comment]
Status report: I restarted firefox today, but the tab hierarchy was restored correctly this time.

To be continued...

**Comment by jarredou** [comment]
Sometimes happens here too... I'll do the debug things described above next time it occurs.

**Comment by artonaut** [comment]
Today, I started firefox using a profile that I hadn't used for quite some time, and the tab tree was partially corrupted. I have attached the debug output, as wel as a screenshot of the tab structure.

[TST-tabs 20230319](https://user-images.githubusercontent.com/121835817/226184645-3594d773-8580-4c47-a3df-43c360f51111.jpg)
[TST-tabs.20230319.txt](https://github.com/piroor/treestyletab/files/11011379/TST-tabs.20230319.txt)

**Comment by artonaut** [comment]
BTW: It looks like I have a profile saved where I can reproduce the problem intermittently (which suggests it is a race condition somewhere...). That should allow me to extract more information, or do additional tests, if you need any.

**Comment by artonaut** [comment]
I did some more tests with the profile I mentioned in my last post. From my testing, it seems that if I press 'Restore Session' as soon as possible after starting Firefox, then often, the tab hierarchy fails to be restored, or is restored only partly. If I wait some time (e.g. 30 seconds) before pressing 'Restore Session', then the tab hierarchy seems to be restored correctly with a high probability, if not every time.

All tests were done as follows:
- Remove current profile directory (`rm -r <directory-name>`)
- Unpack saved archive (`tar -xf <tar-file-of-old-profile-directory>`)
- Start firefox with that profile
- Press 'Restore Session'

The profile from the saved archive dates from (i.e. was last used on) 11 march 2022.

I am attaching the debug output.

This time, I also created an image of the tab hierarchy after it has been (or seems to have been) restored correctly. However, even then, I think that the 'Restore session' tab still has not been restored to its correct place in the hierarchy. This seems to always be the case for the 'Restore Session' tab, even if the rest of the hierarchy is restored correctly.

The files suffixed `-h` were made when the tab hierarchy was restored correctly. The ones suffixed `-f`, are made with a 'flat' tab hierarchy.
[TST-tabs Bm-20230322-f](https://user-images.githubusercontent.com/121835817/227050104-28695984-9c7c-4f17-872c-15e3ca7631f8.jpg)
[TST-tabs.Bm-20230322-f.txt](https://github.com/piroor/treestyletab/files/11045101/TST-tabs.Bm-20230322-f.txt)
[TST-tabs Bm-20230322-h](https://user-images.githubusercontent.com/121835817/227050107-a8a39c8e-1665-4bd0-9168-e9cb7c86e44a.jpg)
[TST-tabs.Bm-20230322-h.txt](https://github.com/piroor/treestyletab/files/11045102/TST-tabs.Bm-20230322-h.txt)

**Comment by artonaut** [comment]
After writing the previous comment, I thought of one additional use-case to test:

Test 1:
- Unpack profile archive
- Start firefox, and restore session as soon as possible. Result: hierarchy not restored
- Kill firefox, and restart without unpacking archive again. Restore session as soon as possible. Each time: hierarchy not restored.

Test 2:
- Unpack profile archive
- Start firefox, and restore session as soon as possible. Result: hierarchy not restored
- Kill firefox, and restart without unpacking archive again. Wait 30 seconds, and restore session. Result: hierarchy restored to its former glory.

**Comment by piroor** [comment]
@artonaut Thank you for details!

> Press 'Restore Session'

Is the button in the `about:sessionrestore` page? You told about tab restoration of the crash recovery feature of Firefox itself? (If the tab has a heading like "Sorry. We’re having trouble getting your pages back." it is the `about:sessionrestore` page.) 

TST's integrated tree restoration on session restoration is basically designed for these triggers:

* Manual restoration: "History" => "Restore previous session"
* Auto restoration: Firefox's options => "General" => "Startup" => Check to "Open previous windows and tabs"

Could you try these cases also? If you don't see this tabs restoration problem, the crash recovery context may be one of required condition to reproduce the problem.

**Comment by artonaut** [comment]
> Is the button in the `about:sessionrestore` page? You told about tab restoration of the crash recovery feature of Firefox itself? (If the tab has a heading like "Sorry. We’re having trouble getting your pages back." it is the `about:sessionrestore` page.)

Yes.
Usually, for me, a firefox restart is not intentional :-). E.g. an OS crash or something of the sort. Then, when I start firefox after booting again, I get the `about:sessionrestore` page. 'Restore Previous Session' does not work at that time...

I did some more tests:

I created an entirely new firefox profile, installed TST (and nothing else), then killed the process, copied the `recovery.jsonlz4` file of the archived profile that fails to restore the tabs, restarted firefox using that new profile, and pressed 'Restore Session'. The tab hierarchy was not restored. That seems to suggest that the problem is (partly) related to something in the saved session ?
I ran the debug code; the output is attached. Noteworthy difference was that this time, that all tabs had their 'ancestors' field empty...
[TST-tabs.Bm-20230331.txt](https://github.com/piroor/treestyletab/files/11116865/TST-tabs.Bm-20230331.txt)

Using the same profile, I started firefox, waited 30s, and restored the session. The hierarchy was restored correctly. I then created a new tab, and opened a web page. I then killed firefox, and restarted it a number of times, but now the tab hierarchy was restored correctly every time...

> Auto restoration: Firefox's options => "General" => "Startup" => Check to "Open previous windows and tabs"

After unpacking the archive, I edited the `prefs.js` file before starting firefox, and added `user_pref("browser.startup.page", 3);`, which enables that option. After starting firefox, the session was restored automatically, but the tab hierarchy was not restored.

> Manual restoration: "History" => "Restore previous session"

After the previous test, I quit firefox (File -> Quit), edited `prefs.js` to remove `user_pref("browser.startup.page", 3);`, copied the original `recovery.jsonlz4` to `sessionstore.jsonlz4` and restarted firefox. The 'Restore previous session' option was now active, and I selected it immediately. The tab hierarchy was not restored. There were no (significant) differences in the debug output, compared to the output in my previous post (TST-tabs.Bm-20230322-f.txt).

My last test for today, came after the observation that, when I start firefox with the TST sidebar active, then immediately after startup, the sidebar is empty. After a short time, it initializes, and then 'Restore Session' tab becomes visible in the sidebar.

If I press 'Restore Session' before the sidebar has initialized, there seems to be a high(er) chance of the hierarchy not being restored correctly. The extent to which this happens seems to depend on the exact contents of the session file. Some session files always restore correctly, some only partly, and some (like the one in the tests described above) seemingly never.
If OTOH, I wait until the sidebar is fully initialized, then there seems to be a high (maybe even 100% ?) chance that the hierarchy _is_ restored correctly. This may also explain the effectiveness of the 30s delay mentioned earlier.

**Comment by piroor** [comment]
Did you see broken tree structure only below the "Restore Session" tab?  I've tried to reproduce the problem and I've actually seen similar problem with a crash recovery.

Before crash / after session restoration ("Restore Session" button is pressed *before* TST is loaded):

* A
  * B
    * C
      * D (active)
        * E
          * F
            * G
              * H
                * I
                  * J
  * K
    * L
      * M
    * N


After session restoration ("Restore Session" button is pressed *while* TST is loading):

* A
  * B
    * C
* D (active) : unexpectedly detached from the tree
  * E
    * F
      * G
        * H
          * I
            * J
* K : unexpectedly detached from the tree
* L : unexpectedly detached from the tree
* M : unexpectedly detached from the tree
* N : unexpectedly detached from the tree


After session restoration ("Restore Session" button is pressed *after for a while*):

* A
  * B
    * C
* D (active) : unexpectedly detached from the tree
  * E
    * F
      * G
        * H
          * I
            * J
  * K: broken structure, this is actually a child of A, not D!
    * L
      * M
    * N

The important point is: broken tree structure always appears below the active tab, on my case.

**Comment by artonaut** [comment]
>Did you see broken tree structure only below the "Restore Session" tab?

No. It also happens that the entire tree hierarchy is broken. As you can see in [one of the screenshots](https://user-images.githubusercontent.com/121835817/227050104-28695984-9c7c-4f17-872c-15e3ca7631f8.jpg) I posted, the complete tree is flat, above and below the 'Restore Session' tab. [The other screenshot](https://user-images.githubusercontent.com/121835817/227050107-a8a39c8e-1665-4bd0-9168-e9cb7c86e44a.jpg) shows how the tab hierarchy is supposed to look, even though the 'Restore Session' tab with its children is still mis-indented: they should not be at the top level, but indented one extra level (IIRC). 

When there are multiple windows, each with many tabs, I remember a case where one and a half window's worth of tabs were restored correctly. whereas the second half of the second window, and all other windows (at least five others), were flat.

Regarding the 'Restore Session' tab, I have the impression that it (with its sub-tabs) ends up mis-indented quite frequently. In addition to that, it is also not getting it's name back quite often: it will still be labeled 'Restore Session' (exactly as you can see in the images), even after a longer time, and even after I've tried to reload it to get the label fixed... Overall, a full or partial failure to restore the complete tab hierarchy of a random saved Firefox session is much less common than just the 'Restore Session' tab being out of line.

> After session restoration ("Restore Session" button is pressed after for a while):
[...]
K: broken structure, this is actually a child of A, not D!

I had a similar situation in my test (see [this image](https://user-images.githubusercontent.com/121835817/227050107-a8a39c8e-1665-4bd0-9168-e9cb7c86e44a.jpg)). Interestingly, in my case, when I collapse the 'Restore Session' tab, the bottom three tabs do not collapse, even though they _seem_ to be children of the 'Restore Session' tab. Conversely, when I collapse the top tab, the bottom three tabs collapse with it... See attached images. I suspect that this will happen in your example as well. That would imply that in your example, `K` is effectively still a child of `A`, and that something special is the case with `D` and its children.
[TST-tabs Bm-20230331-c1](https://user-images.githubusercontent.com/121835817/229225058-a21a91f1-072e-45f0-aacb-d2d5ea57948b.jpg)
[TST-tabs Bm-20230331-c2](https://user-images.githubusercontent.com/121835817/229225059-7921d880-8c47-4b21-b448-9fd2ab5f3540.jpg)


Also, if I open a link in a new tab, from the topmost tab and from the 'Restore Session' tab (so the new tab becomes the bottom-most direct child of the tab it was opened from), the child of the topmost tab ends up at the very bottom of the list of tabs, where it _logically_ belongs, and not above the 'Restore Session' tab, where the hierarchy as shown indicates it would have to go. The child of the 'Restore Session' tab ends up fourth from the bottom, where it belongs _logically_, even though the visual hierarchy would suggest it should end up at the bottom. So the logical structure, even if not shown correctly, still seems to be intact in this case.

I have also tried the same experiment (i.e. opening subtabs) when the entire hierarchy is lost after a restart of Firefox. In that case, the subtabs both end up directly below their parents, suggesting that the logical structure is lost as well, not just the visual hierarchy.

**Comment by artonaut** [comment]
>Did you see broken tree structure only below the "Restore Session" tab?

Or did you mean just the case where the tab hierarchy is restored correctly, except for the 'Restore Session' tab ?

In that case, I misunderstood your question. My apologies. It is probably clear from my comment from yesterday, that in that case, the answer to your question should actually be: yes. If the tree is restored correctly, and the indentation level of the 'Restore Session' tab is not, then, as far as I remember, the tabs above are always fine, and the tabs below are not. Or rather: the tabs below are at their expected indentation levels, but the 'Restore Session' tab having been moved to indentation level 0, they now _seem_ to be children of the 'Restore Session' tab, even though, as I explained yesterday, they don't really behave as such...

**Comment by piroor** [comment]
There are some restrictions around `about:sessionrestore` ("Restore Session") tab due to a bug of Firefox itself.
https://github.com/piroor/treestyletab/issues/2587
https://bugzilla.mozilla.org/show_bug.cgi?id=1640112
The problem you and I saw ("If the tree is restored correctly, and the indentation level of the 'Restore Session' tab is not, then, as far as I remember, the tabs above are always fine, and the tabs below are not") may be a variant of problems from bugs of Firefox itself, then it means that tab restoration after crash recovery mostly fail if the active tab was middle-level in a tree.

**Comment by piroor** [comment]
Sorry for large delay, I've partially understood what happens at the case: https://github.com/piroor/treestyletab/issues/3276#issuecomment-1475320685

![image](https://user-images.githubusercontent.com/70062/231243922-f35b70c7-2c91-4f04-8f7c-f273ebf7282c.png)


TST's tabs should have two-way reference for each: the parent should know its children and children should know its parent. As figured in this screenshot, there are some one-way reference.

* Tabs from 1-3 to 1-23 have correct two-way references. The broken indent is just a visual error, so after you reopen the sidebar or reload TST they should appear correct indentations.
* Relation of tabs 1-14 and 1-15 is partially lost. 1-14 knows which is its children (1-15) but 1-15 does not know its parent, so 1-15 is treated as a top-level tab.
* Relation of 1-24, 1-30 and their children is also lost partially. Only 1-24 and 1-30 know which tabs are their children but their children don't know their parents, so tabs from 1-25 to 1-29 and from 1-31 to 1-35 are treated as top-level tabs.
* Following tabs pair (1-36 and 1-37, 1-38 and 1-39, and more) have one-way reference: their relations are also lost partially.
* All those tabs have restored unique ID provided by TST. This means that all session data were restored by Firefox correctly.

**Comment by piroor** [comment]
Basically TST is designed to save two-way references for all tabs: a child knows which is its parent and a parent knows which tabs are its children. For example, 1-38 should have saved session data "1-39 is my child" and 1-39 also should have "1-38 is my parent". Such two-way references are required to restore tree structure from tabs restored step-by-step. Thus what should happen when tabs are restored is:

* Case A: 1-38 is restored at first and 1-39 is restored later. When 1-38 is restored, TST tries to find its children but 1-39 is not restored yet, so TST cannot restore its tree structure. However, when 1-39 is restored TST can detect 1-38 as its parent, so TST restores tree structure based on the information saved to 1-39. 
* Case B: 1-39 is restored at first and 1-38 is restored later. When 1-39 is restored, TST tries to find its parent but 1-38 is not restored yet, so TST cannot restore its tree structure. However, when 1-38 is restored TST can detect 1-39 as its child, so TST restores tree structure based on the information saved to 1-38. 

Your case looks to be A, but there is a difference: 1-39 did not know which is its parent, but both 1-38 and 1-39 were restored by Firefox correctly. This may mean that TST failed to *save* (not *restore*) the reference from 1-39 to 1-38 at the previous time.

**Comment by piroor** [comment]
For fail-safe, TST also stores one more: complete tree structure information as a part of window's session data. Even if two-way references of each tab were broken, TST can restore complete tree based on the window-related information. But there is an exception: crash recovery. A window recycled for a crash recovery may lose its window-related session information. On such case TST lost last hint to restore tree, and finally you'll get just partially restored tree. I think this is what happened on your case.

**Comment by piroor** [comment]
Here is a script to detect one-way references which can be broken tree on crash recovery:

```javascript
(async () => {
const start = Date.now();
const shouldBeChildren = {};
const shouldBeParent = {};

const restoredInfo = {};

const tabs = Tab.getAllTabs().filter(tab => tab.$TST.parentId || tab.$TST.hasChild);
const lines = await Promise.all(tabs.map(async tab => {
  const [uniqueId, ancestors, children] = await Promise.all([
    browser.sessions.getTabValue(tab.id, 'data-persistent-id'),
    browser.sessions.getTabValue(tab.id, 'ancestors').then(data => data || []),
    browser.sessions.getTabValue(tab.id, 'children').then(data => data || []),
  ]);
  const info = {
    id: tab.id,
    uid: uniqueId.id,
    ancestors,
    children,
  };
  restoredInfo[uniqueId.id] = info;
  return `
---${tab.windowId}-${tab.id}---
id: ${uniqueId.id}
  ancestors: ${ancestors}
  children:  ${children}
`.trim();
}));

for (const [id, info] of Object.entries(restoredInfo)) {
  if (info.ancestors.length > 0) {
    const parent = restoredInfo[info.ancestors[0]];
    if (!parent)
      console.log(`ERROR: ${info.id} should have parent but missing!`);
    else if (!parent.children.includes(info.uid))
      console.log(`ERROR: ${info.id} should be a child of ${parent.id} but the parent does not know ${info.id} !`);
  }
  if (info.children.length > 0) {
    for (const id of info.children) {
      const child = restoredInfo[id];
      if (!child)
        console.log(`ERROR: ${info.id} should have child but missing!`);
      else if (!child.ancestors.includes(info.uid))
        console.log(`ERROR: ${info.id} should be a parent of ${child.id} but the child does not know ${info.id} !`);
    }
  }
}

console.log('elapsed: ', Date.now() - start)

})();
```

and I got:

```
ERROR: 81 should be a parent of 82 but the child does not know 81 !
ERROR: 81 should be a parent of 84 but the child does not know 81 !
ERROR: 82 should be a parent of 83 but the child does not know 82 !
ERROR: 104 should be a parent of 105 but the child does not know 104 !
ERROR: 111 should be a parent of 112 but the child does not know 111 !
ERROR: 120 should be a parent of 121 but the child does not know 120 !
ERROR: 133 should be a parent of 134 but the child does not know 133 !
ERROR: 139 should be a parent of 140 but the child does not know 139 !
ERROR: 139 should be a parent of 141 but the child does not know 139 !
ERROR: 139 should be a parent of 142 but the child does not know 139 !
ERROR: 143 should be a parent of 144 but the child does not know 143 !
ERROR: 164 should be a parent of 165 but the child does not know 164 !
ERROR: 164 should be a parent of 166 but the child does not know 164 !
ERROR: 164 should be a parent of 167 but the child does not know 164 !
ERROR: 164 should be a parent of 168 but the child does not know 164 !
ERROR: 164 should be a parent of 169 but the child does not know 164 !
ERROR: 171 should be a parent of 172 but the child does not know 171 !
ERROR: 171 should be a parent of 175 but the child does not know 171 !
ERROR: 173 should be a parent of 174 but the child does not know 173 !
ERROR: 190 should be a parent of 191 but the child does not know 190 !
ERROR: 198 should be a parent of 199 but the child does not know 198 !
ERROR: 204 should be a parent of 205 but the child does not know 204 !
ERROR: 205 should be a parent of 206 but the child does not know 205 !
ERROR: 209 should be a parent of 210 but the child does not know 209 !
ERROR: 218 should be a parent of 219 but the child does not know 218 !
ERROR: 223 should be a parent of 224 but the child does not know 223 !
ERROR: 223 should be a parent of 225 but the child does not know 223 !
ERROR: 223 should be a parent of 226 but the child does not know 223 !
ERROR: 254 should be a parent of 255 but the child does not know 254 !
ERROR: 255 should be a parent of 256 but the child does not know 255 !
...
```

To be honest I've surprised that it reports such a large number errors on the profile my ordinal use. I should introduce something mechanism to fixup such fragile state aggressively...

**Comment by mYnDstrEAm** [comment]
All trees got flattened with v 3.9.20. I restored a session with multiple windows and many tabs. It flickered again when switching the panel from something else like History to Tree Style Tab to make it load the tabs which I created another issue for earlier. I found the solution to the flickering and no tabs showing to be disabling the addon and enabling it again. After that all tabs are shown but are flattened. When I close a tab that had tabs beneath it and then restore it with ctrl+shift+T, the tree structure is recreated for that specific tab only. I have opened the console but there's no new messages after restarting the TST addon.

What should I do to restore the tree structure?

Okay, the solution was to reopen the browser and restore one window at a time using the addon **Tab Session Manager**.
